src/rgw/rgw_data_sync.cc

```diff
@@ -194,6 +232,50 @@ int RGWReadDataSyncStatusCoroutine::operate()
   return 0;
 }
 
+class RGWReadRecoveringShardsCoroutine : public RGWCoroutine {
+  RGWDataSyncEnv *sync_env;
+  rgw_data_sync_status *sync_status;
+  set<int>& recovering_shards;
+
+  map<int, std::set<std::string>> entries_map;
+  int max_entries{1};
+
+public:
+  RGWReadRecoveringShardsCoroutine(RGWDataSyncEnv *_sync_env,
+                                   rgw_data_sync_status *_status, set<int>& _recovering_shards)
+    : RGWCoroutine(_sync_env->cct), sync_env(_sync_env), sync_status(_status), recovering_shards(_recovering_shards)
+  {}
+  int operate() override;
+};
+
+int RGWReadRecoveringShardsCoroutine::operate()
+{
+  reenter(this) {
+    // read recovering shards
+    using ReadRecoveringShardsCR = RGWReadDataSyncRecoveringShardsCR;
+    yield {
+      set<int> shards;
+      for (int shard_id = 0; shard_id < (int)sync_status->sync_info.num_shards; ++shard_id) {
+        shards.insert(shard_id);
+      }
+      call (new ReadRecoveringShardsCR(sync_env, max_entries, shards, entries_map));
+    }
+    if (retcode < 0) {
+      ldout(sync_env->cct, 4) << "failed to read data sync recovering shards with "
+          << cpp_strerror(retcode) << dendl;
+      return set_cr_error(retcode);
+    }
+
+    for (auto entry : entries_map) {
```

Comment:
use of `auto entry` here will copy each entry by value, and this is expensive for `std::set` - prefer `const auto& entry`